### PR TITLE
SERVICES-2091: rework + index to es

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Reasoning
- the project was outdated
- each new provider had to create a new module
- data was indexed into data-api 

## Proposed Changes
- update project
- make providers as simple classes instead of modules
- index data to Elasticsearch

## How to test
-
-
-

### CODE REVIEW

**Review by gabrielmatei — APPROVED** [review]

File: apps/aggregator/src/elastic/elastic.indexer.service.ts [hunk 1]
```diff
@@ -1,51 +1,51 @@
...
-    let totalBalance = new BigNumber(currentTotalBalance);
-    totalBalance = totalBalance.plus(totalLiquidStaking);
-
+  private async setLiquidStakingValues(address: string, epoch: number, totalLiquidStaking: BigNumber) {
     const indexName = this.getIndexName(epoch);
-    await this.elasticService.setCustomValue(`${indexName}`, address, 'totalBalanceWithStake', totalBalance.toString(10));
-    await this.elasticService.setCustomValue(`${indexName}`, address, 'liquidStaking', totalLiquidStaking.toString(10));
+    //await this.apiService.post()
+
+    const totalLiquidStakingStr = totalLiquidStaking.toString(10);
```
Comment (gabrielmatei) [inline apps/aggregator/src/elastic/elastic.indexer.service.ts:38]: maybe we should change `toString()` to `toFixed()`. It is possible for `toString` from `bignumber.js` to stringify very large numbers in scientific notation

File: apps/aggregator/src/snapshots.service.ts [hunk 2]
```diff
@@ -32,23 +35,38 @@ export class SnapshotsService {
...
+
+      if (await this.isSumOfStakedLessOrEgualToContractStake(providerName, stakedValueSum, await providerInstance.getStakingContracts())) {
+        this.addProviderRecordsToSum(providerUsers, liquidStakingUsers);
+      } else {
+        this.logger.error(`Liquid Staking provider sum of users stake is larger than the staking contracts stake. Provider=${providerName}, Sum=${stakedValueSum.toString(10)}`);
       }
     }
 
+    this.jsonExporter.exportJson(liquidStakingUsers);
     await this.addRecordsToIndexer(liquidStakingUsers);
```
Comment (gabrielmatei) [inline apps/aggregator/src/snapshots.service.ts:64]: also, maybe it would be nice to send a slack alert in case something goes wrong

**Review by gabrielmatei — APPROVED** [review]

File: apps/aggregator/src/elastic/elastic.indexer.service.ts (hunk 1 above)
Comment (bogdan-rosianu) [reply apps/aggregator/src/elastic/elastic.indexer.service.ts:38]: done

File: apps/aggregator/src/snapshots.service.ts [hunk 3]
```diff
@@ -32,23 +35,38 @@ export class SnapshotsService {
...
       } catch (e) {
-        await this.alertsService.sendIndexerError(`Error while indexing data for ${provider}: ${e}`);
-        this.logger.error(`Error while indexing data for ${provider}: ${e}`);
+        await this.alertsService.sendIndexerError(`Error while indexing data for ${providerName}: ${e}`);
+        this.logger.error(`Error while indexing data for ${providerName}: ${e}`);
+        continue;
+      }
+
+
+      if (await this.isSumOfStakedLessOrEgualToContractStake(providerName, stakedValueSum, await providerInstance.getStakingContracts())) {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply apps/aggregator/src/snapshots.service.ts:56]: done

File: apps/aggregator/src/snapshots.service.ts (hunk 2 above)
Comment (bogdan-rosianu) [reply apps/aggregator/src/snapshots.service.ts:64]: done

File: apps/aggregator/src/elastic/elastic.indexer.service.ts [hunk 4]
```diff
@@ -55,4 +55,40 @@ export class ElasticIndexerService {
...
+    await this.apiService.put(`${this.apiConfigService.getElasticUrl()}/${indexName}`, JSON.parse(mapping), this.getApiSettingForESQueries());
+    return true;
+  }
+
+  private async doesIndexExist(indexName: string): Promise<boolean> {
+    try {
+      const result = await this.apiService.head(`${this.apiConfigService.getElasticUrl()}/${indexName}`);
+      return result && result.status && result.status == 200;
+    } catch (error) {
+      return false;
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply apps/aggregator/src/elastic/elastic.indexer.service.ts:93]: done

**Review by gabrielmatei — APPROVED** [review]

File: apps/aggregator/src/elastic/elastic.indexer.service.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      this.logger.warn(`Elasticsearch is not writable: ${error}`);
+      return false;
+    }
+
+    return true;
+  }
+
+  private async setLockedEgldValues(address: string, epoch: number, totalLockedEgld: BigNumber) {
+    const indexName = this.getIndexName(epoch);
+    //await this.apiService.post()
```
Comment (dragos-rebegea) [inline apps/aggregator/src/elastic/elastic.indexer.service.ts:54]: commented code

File: apps/aggregator/src/snapshots.service.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  constructor(
+    private readonly baseProvider: BaseProvider,
+    private readonly apiService: ApiService,
+    private readonly apiConfigService: ApiConfigService,
+    private readonly alertsService: AlertsService,
+    private readonly elasticIndexer: ElasticIndexerService,
+    private readonly jsonExporter: JsonExporterService,
+  ) { }
+
+  @Cron(CronExpression.EVERY_DAY_AT_10AM)
```
Comment (dragos-rebegea) [inline apps/aggregator/src/snapshots.service.ts:26]: can we have a random time so we don`t have users that tries to play the snapshot process in their favor

File: apps/aggregator/src/elastic/elastic.indexer.service.ts (hunk 5 above)
Comment (bogdan-rosianu) [reply apps/aggregator/src/elastic/elastic.indexer.service.ts:54]: fixed

File: apps/aggregator/src/snapshots.service.ts (hunk 6 above)
Comment (bogdan-rosianu) [reply apps/aggregator/src/snapshots.service.ts:26]: currently it's just a placeholder and will be changed when deploying the service

**Review by dragos-rebegea — APPROVED** [review]

**Pull request merged**